### PR TITLE
Kernel: Introduce a new model for writing devicetree drivers

### PR DESCRIPTION
**Pull request by spholz**
This PR should make writing devicetree drivers a lot simpler and keep code duplication minimal.

This PR does *not* depend on #25103. But we will need that PR to actually get the physical addresses from the 'reg' properties.

### CODE REVIEW

**Review by Hendiadyoin1 — APPROVED** [review]
LGTM
(am biased though)

**Review by nico — APPROVED** [review]
Looks good.

File: Userland/Libraries/LibDeviceTree/DeviceTree.cpp [hunk 1]
```diff
@@ -73,4 +73,24 @@ ErrorOr<NonnullOwnPtr<DeviceTree>> DeviceTree::parse(ReadonlyBytes flattened_dev
     return device_tree;
 }
 
+bool DeviceTreeNodeView::is_compatible_with(StringView wanted_compatible) const
```
[the remark this answers is not in the document]
Comment (spholz) [reply Userland/Libraries/LibDeviceTree/DeviceTree.cpp:76]: The spec calls it "a compatible string" or "a compatible value". Not sure where I got that "a compatible" phrase from.
Should I change it to "compatible string"?

File: Kernel/Bus/USB/USBDevice.cpp [hunk 2]
```diff
@@ -31,7 +31,7 @@ ErrorOr<NonnullLockRefPtr<Device>> Device::try_create(USBController& controller,
     // "probe" function, which initialises the local state for the device driver.
     // It is currently the driver's responsibility to search the configuration/interface
     // and take the appropriate action.
```
[the remark this answers is not in the document]
Comment (spholz) [reply Kernel/Bus/USB/USBDevice.cpp:33]: That doesn't fit in the commit title :grimacing:

File: Kernel/Arch/init.cpp [hunk 3]
```diff
@@ -78,6 +78,10 @@
 #    include <Kernel/Arch/riscv64/Delay.h>
 #endif
 
+#if ARCH(AARCH64) || ARCH(RISCV64)
```
[the remark this answers is not in the document]
Comment (spholz) [reply Kernel/Arch/init.cpp:81]: I would probably write some docs once we actually get data from the devicetree nodes, which should also be explained there.

File: Userland/Libraries/LibDeviceTree/DeviceTree.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (nico) [reply Userland/Libraries/LibDeviceTree/DeviceTree.cpp:76]: Anything with correct grammar works :)

File: Kernel/Bus/USB/USBDevice.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (nico) [reply Kernel/Bus/USB/USBDevice.cpp:33]: Is s/Make/Let/ enough or are we 2 over?

File: Userland/Libraries/LibDeviceTree/DeviceTree.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (spholz) [reply Userland/Libraries/LibDeviceTree/DeviceTree.cpp:76]: I also can't think of an expressive commit title that fits in the char limit here.
Or do you think "LibDeviceTree: Add `DeviceTreeNodeView::is_compatible_with`" is good enough?

[the remark this answers is not in the document]
Comment (nico) [reply Userland/Libraries/LibDeviceTree/DeviceTree.cpp]: Sure, that works. (It's also always possible to write more in a paragraph after the first line...)

**Pull request merged**